### DISCUSSION TITLE
Syslog-ng contrib Helm K8S

### DISCUSSION THREAD

**Discussion by rodnymorales5** [syslog-ng/syslog-ng#3979] [General]
I have a question, in the contribution to install syslog-ng in kubernetes via helm, I have a question if you have HPA, how do you manage disk-buffering so as not to lose messages? When a pod has a problem and is deleted or scale by resource consumption, how the following pod is able to resume the sequence to write?

**Comment by bazsi** [comment]
I haven't checked that helm but my assumption is that it would only work with stateless containers, so disk buffers are not really supported with that.

**Reply by rodnymorales5** [reply]
Thanks for the quick reply, maybe @MrAnno  might have some idea, he set up that contribution.

**Comment by bazsi** [comment]
One solution with this use-case that I had in mind recently is to add a kafka based queueing mechanism or at least a kafka source. With that syslog-ng could store its state in kafka, meaning that the syslog-ng container itself could become stateless.

There was some code already for kafka sources but it never materialized. A queue would be a lot nicer.

I don't know how k8s manages state for something like a database, but it could work to put the syslog-ng queue files, plus its persistent state to a shared volume, which could be mounted into containers running syslog-ng, even in cases where a container is migrating to a different node.

**Comment by bazsi** [comment]
This is the description I found on k8s volumes: https://kubernetes.io/docs/concepts/storage/persistent-volumes/

Again, I have never used k8s myself, but by my understanding you should be able to request a volume to be bound into a container and ensure that syslog-ng puts its state on that volume. The question is what happens if the container is migrated to another node, I guess if the backing storage is available to any of the nodes, then that volume would be mounted at the new location allowing syslog-ng to continue where it left off.

In case of automatic scaling, it is possible to provision a new volume, which starts up as empty. The important part is to deprovision auto-scaled syslog-ng instances by first removing it from the load balancer (e.g. to avoid it getting more data), then waiting until the queue empties. Once the queue is empty, you can simply delete the container and the volume with it.

**Comment by Kokan** [comment]
The persistent volume just works as expected, but that only resolves issue when the same container needs to be re-created (like crashed app needs to continue the work). But even if you have shared volume, syslog-ng won't pick up diskq files from other syslog-ng instance.

kubernetes won't solve scaling for the application so the application still have to be stateless or solve it by themself. With a configured disk-buffer file syslog-ng is sure not a stateless application.

As Bazsi mentioned upscaling is non issue, but the down scaling is kinda not possible as of now with disk buffer. Even the suggestion to flush the content of the queue likely won't work, because what if the destination is not available so syslog-ng cannot empty its queue. It does not have a lot of time to do it, as it was asked to scale down for a reason (iirc kubernetes even just kill the container after a while).
But hey, downscaling syslog-ng when there is a non empty diskq file likely not the best idea as there is clearly some work to do.

I would say it would be a good to have a mode in syslog-ng when multiple syslog-ng with the exact destintion could share diskq files between each other. If there are multiple syslog-ng instances configured with the same destination with a disk buffer file, and a syslog-ng instance is terminated, the remaining syslog-ng should pick the the diskq files left by the terminted instance. (of course doing this is far from trivial)
